app/services/blacklight/search_service.rb

```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+# SearchService returns search results from the repository
+module Blacklight
+  class SearchService
+    include Blacklight::RequestBuilders
+
+    def initialize(blacklight_config, user_params = {})
+      @blacklight_config = blacklight_config
+      @user_params = user_params
+    end
+
+    # TODO: Can this be private?
+    attr_reader :blacklight_config
+
+    # a solr query method
+    # @param [Hash] user_params ({}) the user provided parameters (e.g. query, facets, sort, etc)
+    # @yield [search_builder] optional block yields configured SearchBuilder, caller can modify or create new SearchBuilder to be used. Block should return SearchBuilder to be used.
+    # @return [Blacklight::Solr::Response] the solr response object
+    def search_results
+      builder = search_builder.with(user_params)
+      builder.page = user_params[:page] if user_params[:page]
+      builder.rows = (user_params[:per_page] || user_params[:rows]) if user_params[:per_page] || user_params[:rows]
+
+      builder = yield(builder) if block_given?
+      response = repository.search(builder)
+
+      if response.grouped? && grouped_key_for_results
+        [response.group(grouped_key_for_results), []]
+      elsif response.grouped? && response.grouped.length == 1
+        [response.grouped.first, []]
+      else
+        [response, response.documents]
+      end
+    end
+
+    # retrieve a document, given the doc id
+    # @param [Array{#to_s},#to_s] id
+    # @return [Blacklight::Solr::Response, Blacklight::SolrDocument] the solr response object and the first document
+    def fetch(id = nil, extra_controller_params = {})
+      if id.is_a? Array
+        fetch_many(id, extra_controller_params)
+      else
+        fetch_one(id, extra_controller_params)
+      end
+    end
+
+    ##
+    # Get the solr response when retrieving only a single facet field
+    # @return [Blacklight::Solr::Response] the solr response
+    def facet_field_response(facet_field, extra_controller_params = {})
+      query = search_builder.with(user_params).facet(facet_field)
+      repository.search(query.merge(extra_controller_params))
+    end
+
+    # Get the previous and next document from a search result
+    # @return [Blacklight::Solr::Response, Array<Blacklight::SolrDocument>] the solr response and a list of the first and last document
+    def previous_and_next_documents_for_search(index, request_params, extra_controller_params = {})
```

Comment:
I think `previous_and_next_documents_for_search` *might* belong in a totally different service.  I don't like the idea of a `user_params` setter much.  That has too much mutable state for my liking, and it would only be used when invoking this one method.